File: java/src/main/java/com/microsoft/tunnels/contracts/TunnelEndpointStatics.java
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

package com.microsoft.tunnels.contracts;

import java.net.URI;

import org.apache.maven.shared.utils.StringUtils;

class TunnelEndpointStatics {
  public static java.net.URI getPortUri(TunnelEndpoint endpoint, int portNumber) {
    if (portNumber == 0 && !StringUtils.isBlank(endpoint.defaultWebUri)) {
      return URI.create(endpoint.defaultWebUri);
    }

    if (StringUtils.isBlank(endpoint.portUriFormat)) {
      return null;
    }
    return URI.create(endpoint.portUriFormat.replace(
        TunnelEndpoint.portToken, Integer.toString(portNumber)));
  }

  public static String getPortSshCommand(TunnelEndpoint endpoint, int portNumber) {
    if (portNumber == 0 && !StringUtils.isBlank(endpoint.defaultSshCommand)) {
      return endpoint.defaultSshCommand;
    }

    if (StringUtils.isBlank(endpoint.portSshCommandFormat)) {
      return null;
    }
    return endpoint.portSshCommandFormat.replace(
        TunnelEndpoint.portToken, Integer.toString(portNumber));
  }
}

File: ts/src/contracts/tunnelEndpointStatics.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { TunnelEndpoint as ITunnelEndpoint, portToken } from './tunnelEndpoint';

/**
 * Gets a URI where a web client can connect to a tunnel port.
 *
 * Requests to the URI may result in HTTP 307 redirections, so the client may need to
 * follow the redirection in order to connect to the port.
 *
 * If the port is not currently shared via the tunnel, or if a host is not currently
 * connected to the tunnel, then requests to the port URI may result in a 502 Bad Gateway
 * response.
 *
 * @param endpoint The tunnel endpoint containing connection information.
 * @param portNumber The port number to connect to; the port is assumed to be
 * separately shared by a tunnel host.
 * @returns URI for the requested port, or `undefined` if the endpoint does not support
 * web client connections.
 */
export function getPortUri(endpoint: ITunnelEndpoint, portNumber?: number): string | undefined {
    if (!endpoint) {
        throw new TypeError('A tunnel endpoint is required.');
    }

    if (typeof portNumber !== 'number' || !endpoint.portUriFormat) {
        return undefined;
    }

    return endpoint.portUriFormat.replace(portToken, portNumber.toString());
}

export function getPortSshCommand(
    endpoint: ITunnelEndpoint,
    portNumber?: number,
): string | undefined {
    if (!endpoint) {
        throw new TypeError('A tunnel endpoint is required.');
    }

    if (typeof portNumber !== 'number' || !endpoint.portSshCommandFormat) {
        return undefined;
    }

    return endpoint.portSshCommandFormat.replace(portToken, portNumber.toString());
}
